### ISSUE TITLE
[Bug] NaN Propagation in SGLang's Custom FlashAttention Kernel under TP/DP + Padding

### ISSUE THREAD

**Issue by zhuoqing-antgroup** [sgl-project/sglang#18660]
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Describe the bug

When using sgl_kernel.flash_attn.flash_attn_varlen_func, if atten_tp != 1, the following code is used to pad tokens. For example, launch the server with:
launch_server --model-path xxx --tp 4 --dp 2 --enable-dp-attention --enable-dp-lm-head --attention-backend fa3
`global_num_tokens = self.global_num_tokens_cpu
        sync_group_size = len(global_num_tokens)
        attn_tp_size = get_attention_tp_size()
        
    for i in range(sync_group_size):
        # make sure that the padded length is divisible by attn_tp_size because we may need reduce-scatter across attn_tp dim.
        # there is no reduce-scatter in LM logprob, so we do not need to adjust the padded length for logprob
        global_num_tokens[i] = ceil_align(global_num_tokens[i], attn_tp_size)

    dp_padding_mode = DpPaddingMode.get_dp_padding_mode(
        self.is_extend_in_batch, global_num_tokens
    )
    self.dp_padding_mode = dp_padding_mode`
Then we observe a critical numerical stability issue:
● If the input Q/K/V tensors contain trailing NaN tokens (e.g., due to padding or explicit injection), and cu_seqlens_q/k + max_seqlen_q/k are correctly set to exclude those NaN tokens (i.e., reflect only valid length),
the output still becomes partially or fully NaN—including in the valid region.
⚠️ Critical Contrast: The official flash-attn library (from flash_attn import flash_attn_varlen_func) produces clean, NaN-free outputs under identical inputs.
This indicates a bug in SGLang’s custom CUDA kernel implementation—likely related to improper masking of invalid tokens during softmax or accumulation, causing NaN values to propagate into valid positions.
Additionally:
● Output beyond max_seqlen_q is not zeroed out, leaving uninitialized/garbage values.

### Reproduction

Steps to Reproduce 
1. Install SGLang v0.5.8  
2. Run the script below  
3. Observe that recomputed_output contains widespread NaNs, even though only the last token (out of 200) is NaN and max_seqlen_q = 199Expected Behavior
● Valid tokens (positions [0, max_seqlen_q)) should match the output of official flash-attn and remain free of NaNs
● Output beyond max_seqlen_q should be explicitly zeroed (0.0)
Minimal Reproduction Code：
`import torch
from sgl_kernel.flash_attn import flash_attn_varlen_func
import numpy as np

def generate_single_qkv_packed(
    token_length: int,
    head_num: int = 8,
    head_dim: int = 128,
    dtype: torch.dtype = torch.float16,
    device: str = 'cpu',
    nan_last_tokens: int = 0,
    seed: int = 42
):
    if nan_last_tokens < 0:
        raise ValueError("`nan_last_tokens` must be >= 0")
    if nan_last_tokens > token_length:
        raise ValueError(f"`nan_last_tokens` ({nan_last_tokens}) > token_length ({token_length})")

    torch.manual_seed(seed)
    if str(device).startswith('cuda'):
        torch.cuda.manual_seed_all(seed)

    valid_length = token_length - nan_last_tokens

    q = torch.randn((token_length, head_num, head_dim), dtype=dtype, device=device)
    k = torch.randn((token_length, head_num, head_dim), dtype=dtype, device=device)
    v = torch.randn((token_length, head_num, head_dim), dtype=dtype, device=device)

    if nan_last_tokens > 0:
        start_nan = token_length - nan_last_tokens
        q[start_nan:, :, :] = torch.nan
        k[start_nan:, :, :] = torch.nan
        v[start_nan:, :, :] = torch.nan

    cu_seqlens = torch.tensor([0, valid_length], dtype=torch.int32)
    return {
        'q': q,
        'k': k,
        'v': v,
        'cu_seqlens_q': cu_seqlens.clone(),
        'cu_seqlens_k': cu_seqlens.clone(),
        'max_seq_len_q': valid_length,
        'max_seq_len_k': valid_length,
        'total_tokens': token_length,
        'valid_length': valid_length
    }

def safe_tensor_to_float(tensor):
    if tensor.dtype in (torch.bfloat16, torch.float16):
        return tensor.float()
    return tensor

def analyze_token_stats(tensor: torch.Tensor, name: str):
    print(f"\n📊 Analyzing {name} (shape: {tensor.shape})")
    print("-" * 80)
    print(f"{'Token':<6} {'Mean':<10} {'Median':<10} {'Std':<10} {'Min':<10} {'Max':<10} {'NaN':<6} {'Inf':<6}")
    print("-" * 80)

    tensor_f32 = safe_tensor_to_float(tensor)
    T, H, D = tensor_f32.shape

    for t in range(T):
        token = tensor_f32[t]
        flat = token.flatten()

        nan_mask = torch.isnan(flat)
        inf_mask = torch.isinf(flat)
        valid_mask = ~(nan_mask | inf_mask)
        num_nan = nan_mask.sum().item()
        num_inf = inf_mask.sum().item()

        if valid_mask.any():
            valid_vals = flat[valid_mask]
            mean_val = valid_vals.mean().item()
            median_val = valid_vals.median().item()
            std_val = valid_vals.std().item()
            min_val = valid_vals.min().item()
            max_val = valid_vals.max().item()
        else:
            mean_val = median_val = std_val = min_val = max_val = float('nan')

        def fmt(x):
            if np.isnan(x):
                return "NaN"
            elif np.isinf(x):
                return "Inf" if x > 0 else "-Inf"
            elif abs(x) > 1e4:
                return f"{x:.2e}"
            else:
                return f"{x:.4f}"

        print(
            f"{t:<6} "
            f"{fmt(mean_val):<10} "
            f"{fmt(median_val):<10} "
            f"{fmt(std_val):<10} "
            f"{fmt(min_val):<10} "
            f"{fmt(max_val):<10} "
            f"{num_nan:<6} "
            f"{num_inf:<6}"
        )

# Generate input with 1 trailing NaN
saved_inputs = generate_single_qkv_packed(
    token_length=200,
    head_num=8,
    head_dim=128,
    nan_last_tokens=1,
    dtype=torch.bfloat16,
    device='cuda',
    seed=42
)

device = "cuda"
for key in ["q", "k", "v"]:
    saved_inputs[key] = saved_inputs[key].to(device=device)

for key in ["cu_seqlens_q", "cu_seqlens_k"]:
    saved_inputs[key] = saved_inputs[key].to(device=device, dtype=torch.int32)

saved_inputs["max_seqlen_q"] = int(saved_inputs["max_seq_len_q"])
saved_inputs["max_seqlen_k"] = int(saved_inputs["max_seq_len_q"])

print('max_seqlen_q:', saved_inputs["max_seqlen_q"])
print('cu_seqlens_q:', saved_inputs["cu_seqlens_q"])
print('Number of NaNs in q:', torch.isnan(saved_inputs["q"]).sum().item())

# Run SGLang's FlashAttention
try:
    recomputed_output = flash_attn_varlen_func(
        q=saved_inputs["q"],
        k=saved_inputs["k"],
        v=saved_inputs["v"],
        cu_seqlens_q=saved_inputs["cu_seqlens_q"],
        cu_seqlens_k=saved_inputs["cu_seqlens_k"],
        max_seqlen_q=saved_inputs["max_seqlen_q"],
        max_seqlen_k=saved_inputs["max_seqlen_k"],
        causal=True,
    )
    analyze_token_stats(recomputed_output, "recomputed_output (SGLang)")
except Exception as e:
    print("❌ Error during recomputation:")
    raise e`

### Environment

● SGLang version: v0.5.8  
● Python version: 3.12.3  
● PyTorch version: 2.9.1+cu130  
● CUDA version: 13.0  
● OS: Ubuntu 24.04.2 LTS  
● GPU: H20
